backend/app/models.py

```diff
@@ -0,0 +1,144 @@
+import enum
+
+from sqlalchemy.ext.hybrid import hybrid_property
+from werkzeug.security import generate_password_hash, check_password_hash
+
+from init_app import db
+
+
+class Role(enum.Enum):
+    ADMIN = "Admin"
+    SELLER = "Seller"
+    BUYER = "Buyer"
+
+
+class User(db.Model):
+    id = db.Column(db.Integer, primary_key=True)
+    username = db.Column(db.String(80), unique=True, nullable=False)
+    email = db.Column(db.String(120), unique=True, nullable=True)
+    _password_hash = db.Column(db.String(128))
+    sold_item_id = db.Column(db.Integer, db.ForeignKey('item.id'))
+    bought_item_id = db.Column(db.Integer, db.ForeignKey('item.id'))
+    sold_item = db.relationship('Item', backref='seller', lazy=True, foreign_keys=[sold_item_id])
+    bought_item = db.relationship('Item', backref='buyer', lazy=True, foreign_keys=[bought_item_id])
+    role = db.Column(db.Enum(Role))
+
+    def __repr__(self):
+        return '<User %r>' % self.username
+
+    @hybrid_property
+    def password_hash(self):
+        return self._password_hash;
+
+    @password_hash.setter
+    def password_hash(self, password):
+        self._password_hash = generate_password_hash(password)
+
+    def check_password(self, password):
+        return check_password_hash(self._password_hash, password)
+
+
+class Item(db.Model):
+    id = db.Column(db.Integer, primary_key=True)
+    name = db.Column(db.String(80), unique=True, nullable=False)
+    _category = db.relationship('Category', backref='item', lazy=True)
+    category_id = db.Column(db.Integer, db.ForeignKey('category.id'), nullable=True)
+    _subcategory = db.relationship('Subcategory', backref='item', lazy=True)
+    subcategory_id = db.Column(db.Integer, db.ForeignKey('subcategory.id'), nullable=True)
+    _image = db.relationship('Image', backref='item', lazy=True)
+
+    def __repr__(self):
+        return '<Item %r>' % self.name
+
+    @hybrid_property
+    def category(self):
+        return self._category
+
+    @category.setter
+    def category(self, category_name):
+        category = Category.query.filter_by(name=category_name).first_or_404()
+        self.category_id = category.id
+
+    @hybrid_property
+    def subcategory(self):
+        return self._subcategory
+
+    @subcategory.setter
+    def subcategory(self, subcategory_name):
+        subcategory = Subcategory.query.filter_by(name=subcategory_name).first_or_404()
+        self.subcategory_id = subcategory.id
+
+    @hybrid_property
+    def image(self):
+        return self._image
+
+    @image.setter
+    def image(self, image_name):
+        image = Image.query.filter_by(name=image_name).first_or_404()
```

Comment:
Wy you query inside a model, you can assume to provide model before setting the attributes, so whole query is done before changes, like we do it in Django

Reply:
I guess the relation should be like one item can have one image, but you can use image with many items

Reply:
Then it destruct idea of having everything which depends on model in models.py -> that's why I write method update_item etc. To make query earlier I have to check parameters earlier (in resources). I can change it but in my opinion it's worse than present one.
In my opinion relation should be opposite because you may want to have more images than one how item looks like.